subsys/net/ip/tcp.c

```diff
@@ -3065,12 +3065,18 @@ int net_tcp_connect(struct net_context *context,
 		 * net_context_create_ipv4_new() is not called in the packet
 		 * output chain
 		 */
-		ip4 = net_if_ipv4_select_src_addr(
-			net_context_get_iface(context),
-			&net_sin(remote_addr)->sin_addr);
-		conn->src.sin.sin_addr = *ip4;
+		if (net_ipv4_is_addr_unspecified(
+			&net_sin(local_addr)->sin_addr)) {
+			ip4 = net_if_ipv4_select_src_addr(
+					net_context_get_iface(context),
+					&net_sin(remote_addr)->sin_addr);
+			net_ipaddr_copy(&conn->src.sin.sin_addr, ip4);
+		} else {
+			net_ipaddr_copy(&conn->src.sin.sin_addr,
+							&net_sin(local_addr)->sin_addr);
```

Comment:
This looks like editor miscofiguration on your side (note that Zephyr uses 8spaces/tab)